### PR TITLE
feat: redux logging middleware with feature flag

### LINKED ISSUE #1226: Add redux logging middleware
### Overview

We heavily use redux for app state, but our store is so large we're crashing the redux dev tools browser extension.

We want a way to be able to dig into state changes via console logs in the browser to debug issues locally.

### Reference

Use some sort of middleware as per the docs https://redux.js.org/tutorials/fundamentals/part-4-store#middleware

### Acceptance Criteria

* Should only log if an environment variable is set - `REACT_APP_REDUX_LOGGING`
* Logging should include action and new state on separate lines
* Logging should be formatted nicely, with expandable objects rather than fully serialized messages
* Should be available in feature flags menu too

### Need By Date

_No response_

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by afenton90**
## Description

Enables Redux logging with the following criteria:

- Should only log if an environment variable is set - `REACT_APP_REDUX_LOGGING`
- Logging should include action and new state on separate lines
- Logging should be formatted nicely, with expandable objects rather than fully serialized messages
- Should be available in feature flags menu too

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1226 

## Risk

Middleware added to Redux store. Minimal risk.

## Testing

Tested in wallet connect, dashboard, and feature flag screens.
With feature flag initially set to false or true. Also togging in the flags menu.

## Screenshots (if applicable)
image

### CODE REVIEW

File: src/state/middleware/logging.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { Middleware } from '@reduxjs/toolkit'
+
+const log = (message: string, obj: unknown, featureFlag: boolean) => {
+  if (featureFlag) {
+    // eslint-disable-next-line no-console
+    console.log(message, obj)
```
Comment (DaoDev44) [inline src/state/middleware/logging.ts:6]: I believe we're trying to default to `console.info` instead of using `console.log`

Comment (afenton90) [reply src/state/middleware/logging.ts]: Thanks I've updated to info now

**Review by DaoDev44 — APPROVED** [review]

File: src/state/middleware/logging.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Middleware } from '@reduxjs/toolkit'
+
+const log = (message: string, obj: unknown, featureFlag: boolean) => {
```
Comment (gomesalexandre) [inline src/state/middleware/logging.ts:3]:
```suggestion
const log = (message: string, obj: Record<string, unknown>, featureFlag: boolean) => {
```

**Review by gomesalexandre — COMMENTED** [review]
Thanks for your contribution @afenton90, great work!

Tested and it works great:
- Logging is disabled when starting the app without the toggle in env/with toggle disabled
- Logging is enabled when starting the app with the toggle enabled in env
- Logging stops when disabling the flag and is re-enabled when re-enabling it 

- Action dispatches are logged as unserialized objects with the ability to be expanded
- New states are logged as unserialized objects with the ability to be expanded

Before accepting this, I'm wondering, would LogRocket's logger work for us with feature flags https://github.com/LogRocket/redux-logger?
It would allow us not to reinvent the wheel and displays the prevState, action and newState as one object so state transitions are more easily identifiable.

Perfectly happy with this solution if it doesn't.

**Comment by afenton90** [comment]
Thanks @gomesalexandre glad its working well for you!
On the addition of LogRockets `redux-logger` I guess it depends on how complex you want to go with the logging format in this project? The pros of having the middleware as it is are that it's small and within your gift to control behaviour and the output.
However, if the requirements for logging redux state expand beyond the current spec, I could see a case for bringing in a 3rd party module such as `redux-logger` for added features.

**Comment by 0xdef1cafe** [comment]
> 

hey @afenton90 - i actually agree with @gomesalexandre here, logrocket's solution looks nice and extensible.

do you mind implementing that rather than a hand rolled solution? we can absolutely adjust the bounty price for the additional work, just let us know how long you spent :)

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
requesting logrocket middleware as per @gomesalexandre suggestion

**Comment by afenton90** [comment]
@0xdef1cafe @gomesalexandre that's not a problem 👍 I've added in `redux-logger` as the middleware which will still be associated with the feature flag so it all still works. That's all pushed up to the branch now.
Only spent another half hour doing this, but adding a little bit onto the bounty would be great thanks!

File: src/state/middleware/logging.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { createLogger } from 'redux-logger'
+
+export const logging = createLogger({
+  predicate: getState => getState()?.preferences?.featureFlags?.ReduxLogging
```
Comment (gomesalexandre) [inline src/state/middleware/logging.ts:4]:
```suggestion
  predicate: (getState: () => ReduxState) => getState()?.preferences?.featureFlags?.ReduxLogging

```

**Review by gomesalexandre — APPROVED** [review]
LGTM! Tested again with the same flow as the previous comment and it looks sweet.

Great job @afenton90  🎉

File: src/state/store.ts [hunk 4]
```diff
@@ -18,7 +19,7 @@ const persistConfig = {
 
 registerSelectors(portfolioSelectors)
 
-const apiMiddleware = [portfolioApi.middleware, marketApi.middleware, assetApi.middleware]
+const apiMiddleware = [portfolioApi.middleware, marketApi.middleware, assetApi.middleware, logging]
```
Comment (gomesalexandre) [inline src/state/store.ts:22]: My pedantic self as well as my love for the OCD fam would call that one `loggingMiddleware` but that's just nit-picking

**Review by 0xdef1cafe — APPROVED** [review]

**Review by cjthompson — APPROVED** [review]

**Comment by 0xdef1cafe** [comment]
@afenton90 this is fucking fantastic thank you. are you interested in making a small follow up PR to remove the previous state, and add the diff to the logging? the diff between previous and new state is usually the most useful thing when debugging.

**Comment by afenton90** [comment]
@0xdef1cafe glad to here you like it!
I've done that small follow up PR here https://github.com/shapeshift/web/pull/1288 if you've anymore web app bounties let me know. 👍

**Pull request merged**